File: worker/app/blueprint_contracts/review_flags.py
"""Review flag contract twin (Phase 1)."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal, Optional

ReviewFlagSeverity = Literal["info", "warning", "error", "blocker"]

ReviewFlagRelatedEntityType = Literal[
    "import_session",
    "source_document",
    "detected_trade",
    "accepted_trade",
    "measurement_object",
    "template_binding",
    "material_draft_line",
    "labor_draft_line",
    "plan_path",
]

BLOCKING_FLAG_CODES: frozenset[str] = frozenset(
    {
        "missing_required_measurement",
        "unsupported_trade_for_mvp",
        "paint_without_wall_source",
        "windows_doors_selected_as_trade",
        "missing_plan_path",
        "future_trade_requires_sheet_intelligence",
        "formula_input_missing",
        "template_required_assumption_missing",
    }
)


@dataclass
class BlueprintReviewFlag:
    import_session_id: str
    related_entity_type: ReviewFlagRelatedEntityType
    flag_code: str
    message: str
    severity: ReviewFlagSeverity = "warning"
    blocking: bool = False
    related_entity_id: Optional[str] = None
    resolved: bool = False
    resolved_by: Optional[str] = None
    resolved_at: Optional[str] = None
    metadata: dict[str, Any] = field(default_factory=dict)
    id: Optional[str] = None


def create_review_flag(
    *,
    import_session_id: str,
    related_entity_type: ReviewFlagRelatedEntityType,
    flag_code: str,
    message: str,
    related_entity_id: Optional[str] = None,
    severity: Optional[ReviewFlagSeverity] = None,
    blocking: Optional[bool] = None,
    metadata: Optional[dict[str, Any]] = None,
) -> BlueprintReviewFlag:
    inferred_blocking = blocking if blocking is not None else flag_code in BLOCKING_FLAG_CODES
    inferred_severity: ReviewFlagSeverity = severity if severity is not None else (
        "blocker" if inferred_blocking else "warning"
    )
    return BlueprintReviewFlag(
        import_session_id=import_session_id,
        related_entity_type=related_entity_type,
        flag_code=flag_code,
        message=message,
        severity=inferred_severity,
        blocking=inferred_blocking,
        related_entity_id=related_entity_id,
        metadata=metadata or {},
    )

File: supabase/functions/_shared/blueprint-importer/review-flags.ts
// Blueprint Importer v2 — Review flag contract (Phase 1).
// Side-effect-free.

export type ReviewFlagSeverity = "info" | "warning" | "error" | "blocker";

export type ReviewFlagRelatedEntityType =
  | "import_session"
  | "source_document"
  | "detected_trade"
  | "accepted_trade"
  | "measurement_object"
  | "template_binding"
  | "material_draft_line"
  | "labor_draft_line"
  | "plan_path";

export type ReviewFlagCode =
  | "missing_required_measurement"
  | "unsupported_trade_for_mvp"
  | "paint_without_wall_source"
  | "windows_doors_selected_as_trade"
  | "missing_plan_path"
  | "low_confidence_measurement"
  | "future_trade_requires_sheet_intelligence"
  | "formula_input_missing"
  | "template_required_assumption_missing"
  // non-blocking informational codes are allowed too; consumers should treat
  // unknown codes as warnings by default
  | (string & {});

export interface BlueprintReviewFlag {
  id?: string;
  import_session_id: string;
  related_entity_type: ReviewFlagRelatedEntityType;
  related_entity_id?: string | null;
  severity: ReviewFlagSeverity;
  flag_code: ReviewFlagCode;
  message: string;
  blocking: boolean;
  resolved?: boolean;
  resolved_by?: string | null;
  resolved_at?: string | null;
  metadata?: Record<string, unknown>;
}

const BLOCKING_FLAG_CODES = new Set<string>([
  "missing_required_measurement",
  "unsupported_trade_for_mvp",
  "paint_without_wall_source",
  "windows_doors_selected_as_trade",
  "missing_plan_path",
  "future_trade_requires_sheet_intelligence",
  "formula_input_missing",
  "template_required_assumption_missing",
]);

export function createReviewFlag(input: Omit<BlueprintReviewFlag, "blocking" | "severity"> & {
  severity?: ReviewFlagSeverity;
  blocking?: boolean;
}): BlueprintReviewFlag {
  const inferredBlocking = input.blocking ?? BLOCKING_FLAG_CODES.has(String(input.flag_code));
  const inferredSeverity: ReviewFlagSeverity =
    input.severity ?? (inferredBlocking ? "blocker" : "warning");
  return {
    ...input,
    blocking: inferredBlocking,
    severity: inferredSeverity,
    resolved: input.resolved ?? false,
  };
}
